src/V3Udp.cpp

```diff
@@ -0,0 +1,204 @@
+
+#include "V3PchAstNoMT.h"  // VL_MT_DISABLED_CODE_UNIT
+
+#include "V3Udp.h"
+
+#include <vector>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+//V3Udp add the support for UDP table.
+// For example:
+// table
+//    x 0 1  :   1;
+//    0 ? 1  :   1;
+//    0 1 0  :   0;
+// endtable
+// For every table line, for the input field,
+// two number (mask number and compare number) will
+//  be generated to help make a judegement whether
+// the input field condition is satisfied. For example,
+// for line x 0 1 : 1, mask = 011 cmp = 001, the condition
+// is mask & inputvar == cmp. This passed should be added
+// before V3Inline and V3Tristate.
+
+class UdpVisitor final : public VNVisitor {
+    AstVar* m_ifieldVarp = nullptr;  // input field var of table line.
+    AstVar* m_ofieldVarp = nullptr;  // output filed var of table line.
+    std::vector<AstVar*> m_inputVars;  // All the input vars in the AstPrimitive.
+    std::vector<AstVar*> m_outputVars;  // All the output vars in the AstPrimitive.
+    AstPrimitive* m_primp = nullptr;
+    AstIf* m_lineStmtp = nullptr;  // stmt for every line in UDP Table.
+    AstAlways* m_alwaysp = nullptr;  // UPD Table is realized under the always_latch.
+    bool m_isFirstOutput = false;  // Whether the first IO port is output.
+    int m_inputNum = 0;
+    int m_outputNum = 0;
+
+    void visit(AstPrimitive* nodep) override {
+        m_primp = nodep;
+        m_isFirstOutput = false;
+        m_inputVars.clear();
+        m_outputVars.clear();
+        iterateChildren(nodep);
+        m_primp = nullptr;
+    }
+    void visit(AstVar* nodep) override {
+        // Push the input and output vars for primitive.
+        if (m_primp) {
+            if (nodep->isIO()) {
+                if (nodep->isInput()) {
+                    m_inputVars.push_back(nodep);
+                } else {
+                    m_outputVars.push_back(nodep);
+                }
+                if ((m_inputVars.size() == 0) && (m_outputVars.size() == 1)) {
+                    m_isFirstOutput = true;
+                }
+            }
+        }
+        iterateChildren(nodep);
+    }
+    void visit(AstUdpTable* nodep) override {
+        auto fl = nodep->fileline();
+        m_lineStmtp = nullptr;
+        m_inputNum = m_inputVars.size();
+        m_outputNum = m_outputVars.size();
+        if (m_outputNum != 1) {
+            m_outputVars.back()->v3error(
+                m_outputNum << " output ports for udp table, there must be one output port!");
+        }
+        if (!m_isFirstOutput && m_outputNum) {
+            m_inputVars[0]->v3error("The first port must be the output port!");
+        }
+        m_ofieldVarp = m_outputVars[0];
+        AstBasicDType* const bdtypep = VN_CAST(m_ofieldVarp->childDTypep(), BasicDType);
+        if (bdtypep && bdtypep->isLogic()) {  // If output is reg.
+            bdtypep->v3error("sequetial UDP is not suppoted currently!");
+        }
```

Comment:
Instead test for sequential by looking for an extra divider on a line per IEEE.

Sequential UDPs need to still be UNSUPPORTED warning and suppressed with --bbox (like in the old test).